### PR TITLE
Implement support for collections of collections marshalling

### PR DESCRIPTION
**Pull request by jkoritzinsky**
The current code in the feature/DllImportGenerator branch almost supports passing collections of collections (like jagged arrays) by value or by-ref `in` (it only needs the first commit that handles nested for loops), but it does not correctly support passing collections of collections in a way that requires unmarshalling.

This PR adds the functionality required to enable marshalling collections of collections like jagged arrays from native code back to managed code.

This code is designed such that if n levels of indirection is supported, then n+1 levels is also supported. The `CollectionsOfCollectionsStress` code snippet test validates this by testing 0-11 dimensional jagged arrays.

To enable safely tracing dependencies of "CountElementName" metadata values, this PR enhances the cycle-breaking in the MarshallingAttributeInfoParser to correctly break cycles.

To identify dependencies between parameters/return values, this PR also introduces a topological sorting algorithm to order the parameter/return value unmarshalling in an order that ensures dependencies are marshalled before elements that depend on them.

This algorithm is specified generically so developers of interop source generators could use this mechanism to support their own scenarios. For example, WinRT's implicit array length parameters could have part of their implementation supported by the topological sort.

### CODE REVIEW

File: DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs [hunk 1]
```diff
@@ -46,22 +45,41 @@ internal sealed class ContiguousCollectionElementMarshallingCodeContext : StubCo
...
+                {
+                    i++;
+                }
+                parentContext = context.ParentContext;
+            }
+
+            // Follow a progression of indexers of the following form:
+            // __i, __j, __k, __ii, __jj, __kk, etc.
+            StringBuilder indexerIdentifier = new StringBuilder("__");
+            indexerIdentifier.Append((char)('i' + (i % 3)), 1 + (i / 3));
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs]: I think I would prefer to simply append the depth - `__i0`, `__i1`, ...,  `__iN`, etc. Computing the depth from the current form appears much harder. Is there a benefit of current transform?

File: DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs [hunk 2]
```diff
@@ -12,14 +12,14 @@ namespace Microsoft.Interop
...
     {
-        private readonly string indexerIdentifier;
         private readonly string nativeSpanIdentifier;
-        private readonly StubCodeContext parentContext;
 
         public override bool SingleFrameSpansNativeContext => false;
 
         public override bool AdditionalTemporaryStateLivesAcrossStages => false;
 
+        public string IndexerIdentifier { get; }
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs:21]: `init;`?

File: DllImportGenerator/DllImportGenerator/Marshalling/ICustomNativeTypeMarshallingStrategy.cs [hunk 3]
```diff
@@ -127,26 +127,24 @@ public IEnumerable<StatementSyntax> GeneratePinStatements(TypePositionInfo info,
...
-
         public CustomNativeTypeWithValuePropertyStubContext(StubCodeContext parentContext)
         {
-            this.parentContext = parentContext;
+            ParentContext = parentContext;
             CurrentStage = parentContext.CurrentStage;
         }
 
-        public override bool SingleFrameSpansNativeContext => parentContext.SingleFrameSpansNativeContext;
+        public override bool SingleFrameSpansNativeContext => ParentContext!.SingleFrameSpansNativeContext;
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/ICustomNativeTypeMarshallingStrategy.cs:136]: I don't see the `?` - is the `!` needed?

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs [hunk 4]
```diff
@@ -86,10 +88,152 @@ public static TypeSyntax GetCompatibleGenericTypeParameterSyntax(this TypeSyntax
...
+        /// <returns>A topologically sorted collection of the elemens of the <paramref name="elements"/> collection.</returns>
+        /// <exception cref="InvalidOperationException">The graph of <paramref name="elements"/> nodes and the edges produced by <paramref name="getDependentIndicesFn"/> has cycles.</exception>
+        public static IEnumerable<T> GetTopologicallySortedElements<T, U>(
+            ICollection<T> elements,
+            Func<T, U> keyFn,
+            Func<T, IEnumerable<U>> getDependentIndicesFn)
+        {
+            Dictionary<U, int> elementIndexToEdgeMapNodeId = new(elements.Count);
+            List<T> nodeIdToElement = new(elements.Count);
+            EdgeMap edgeMap = new EdgeMap(elements.Count);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs]: Let's be consistent here with `new` if possible.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs [hunk 5]
```diff
@@ -86,10 +88,152 @@ public static TypeSyntax GetCompatibleGenericTypeParameterSyntax(this TypeSyntax
...
+                    {
+                        S.Add(nodeIdToElement[m]);
+                    }
+                }
+            }
+
+            // If we have edges left, then we have a cycle.
+            if (edgeMap.AnyEdges)
+            {
+                throw new InvalidOperationException();
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs]: Let's add a nice message here. This should be considered a catastrophic error for the source generator.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 6]
```diff
@@ -392,21 +392,72 @@ ExpressionSyntax GetExpressionForParam(TypePositionInfo? paramInfo)
...
+                    {
+                        indexerStack.Push(collectionContext.IndexerIdentifier);
+                    }
+                    lastContext = currentContext;
+                    currentContext = currentContext.ParentContext;
+                }
+
+                numIndirectionLevels = indexerStack.Count;
+
+                ExpressionSyntax indexedNumElements = IdentifierName(lastContext!.GetIdentifiers(numElementsInfo).managed);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs]: Why is the `!` needed here?

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 7]
```diff
@@ -392,21 +392,72 @@ ExpressionSyntax GetExpressionForParam(TypePositionInfo? paramInfo)
...
+                while (indexerStack.Count > 0)
+                {
+                    indexedNumElements = ElementAccessExpression(indexedNumElements).AddArgumentListArguments(Argument(IdentifierName(indexerStack.Pop())));
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs]: Please break this nesting up.

File: DllImportGenerator/DllImportGenerator/StubCodeGenerator.cs [hunk 8]
```diff
@@ -76,6 +77,16 @@ internal sealed class StubCodeGenerator : StubCodeContext
             // Get marshaller for return
             this.retMarshaller = CreateGenerator(retTypeInfo);
 
+
+            List<(TypePositionInfo TypeInfo, IMarshallingGenerator Generator)> allMarshallers = new(this.paramMarshallers);
+            allMarshallers.Add(retMarshaller);
+
+            this.sortedMarshallers = MarshallerHelpers.GetTopologicallySortedElements(
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/StubCodeGenerator.cs:106]: Comment why we are doing this with an example.

File: DllImportGenerator/DllImportGenerator/StubCodeGenerator.cs [hunk 9]
```diff
@@ -202,43 +235,57 @@ public BlockSyntax GenerateSyntax()
...
+                            }
+                            else
+                            {
+                                statements.AddRange(generatedStatements);
+                            }
                         }
                     }
                 }
 
+
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/StubCodeGenerator.cs]: Extra line.

File: DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs (hunk 1 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs]: I like the idea of appending the depth. That's a better idea. I'll switch it to that.

File: DllImportGenerator/DllImportGenerator/Marshalling/ICustomNativeTypeMarshallingStrategy.cs (hunk 3 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/ICustomNativeTypeMarshallingStrategy.cs:136]: `ParentContext` is nullable, so yes this is needed.

Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/Marshalling/ICustomNativeTypeMarshallingStrategy.cs:136]: Oh casing.

File: DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs (hunk 2 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs:21]: This property is calculated internally, so we don't want external code that creates `ContiguousCollectionElementMarshallingCodeContext` instances to assign it.

Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/ContiguousCollectionElementMarshallingCodeContext.cs:21]: Oh. I didn't realize that `init` was also subject to scope - I thought it was limited to constructor.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**